### ISSUE TITLE
[Azure] Is Azure CLI Auth Supported?

### ISSUE THREAD

**Issue by Evesy** [mozilla/sops#418]
Hi,

The docs don't seem to suggest it is but thought I'd double check -- Does sops support Azure CLI login as an authentication method, making use of the generated `~/.azure/accessTokens.json` file to provide auth rather than Client ID/Client Secret?

It would be great for users to be able to encrypt/decrypt using sops with their normal user account (Which is tied to groups to manage permissions, and provides auditability). Needing a service principal for user access means either creating a service principal per user or sharing a service principal amongst users, neither of which seem ideal (As I understand it, very new to Azure).

Any pointers if it is possible would be much appreciated!

Cheers

**Comment by autrilla** [comment]
I don't think this is supported. I have no idea if there's a technical reason for it, and if there's not, we should definitely support it.

**Comment by Evesy** [comment]
It looks like the currently vendored `azure-sdk-for-go` is a few versions behind. The later versions include a number of new methods that I believe would support using CLI auth; `NewAuthorizerFromCLI` https://github.com/Azure/azure-sdk-for-go/blob/master/services/keyvault/auth/auth.go

If this is the case, perhaps it would be possible to add the CLI as a fallback method for auth if the auth from Environment fails (https://github.com/mozilla/sops/blob/master/azkv/keysource.go#L38)

Very new to Golang so the above may be way off, but if it looks like it might be a relatively minor change I'd be happy to try and give it a go. 

Cheers!

**Comment by autrilla** [comment]
Oh, yes, that might have been the reason it wasn't implemented this way in the first place. It would be awesome to support it and I'd be happy to review a PR for it!

**Comment by Evesy** [comment]
Hey @autrilla 

I've unfortunately failed at the first hurdle trying to figure out Go dependencies. Trying to run `govend -u` yields a few errors:

```
go.mozilla.org/yaml.v2 bad ping: unrecognized import path "go.mozilla.org/yaml.v2" (parse https://go.mozilla.org/yaml.v2?go-get=1: no go-import meta tags)
math/bits bad ping: unrecognized import path "math/bits" (import path does not begin with hostname)
```

So I thought I'd just try and update the specific package I'm interested in (`azure-sdk-for-go`) with the below, however I end up in a seemingly infinite chain of required package updates:

```
rm -Rf ./vendor/github.com/Azure/azure-sdk-for-go
govend github.com/Azure/azure-sdk-for-go
```

Am I missing a trick here?

**Comment by autrilla** [comment]
`govend -u github.com/Azure/azure-sdk-for-go` seems to work for me on master.

**Comment by Evesy** [comment]
`govend` is successful however if you run `make test` afterwards should get the below:

```
vendor/github.com/Azure/azure-sdk-for-go/services/keyvault/2016-10-01/keyvault/client.go:28:2: cannot find package "github.com/Azure/go-autorest/tracing" in any of:
```

I can then update the `go-autorest` package to fix that, but it then depends on some new open census packages. I'm not familiar with Golangs package system, is there a way to transitively update all packages that a certain package is dependent on?

**Comment by asw101** [comment]
Hi @Evesy, @autrilla,

I have been working on a PR which adds Azure CLI authentication (as the default), whilst maintaining other authentication methods, and leaving open the possibility of Device Code authentication flow in the future. It also updates Azure/azure-sdk-for-go, Azure/go-autorest and adds the OpenCensus dependencies.

I hope to send a WIP PR shortly, but feel free to take a brief look at my fork at: https://github.com/asw101/sops

**Comment by autrilla** [comment]
@asw101 awesome, thanks for working on that! The code itself looks pretty good to me, but I can't really comment on whether this is the right way to do things since I'm not familiar with Azure at all. If you do submit a PR, it'd be good to get feedback from people who actually use Azure as well as from me.

**Comment by Evesy** [comment]
I don't have strong opinions either way, however, I think it'd be good to keep the order of precedence with sops' Azure authentication in line with the likes of the Azure CLI itself (https://docs.microsoft.com/en-us/cli/azure/azure-cli-configuration?view=azure-cli-latest)

In this case, I'd expect the `AZURE_`environment variables to be preferred over any credentials in `~/.azure`.

**Comment by asw101** [comment]
Thank you for your help @autrilla! I have submitted the PR on https://github.com/mozilla/sops/pull/425.

@autrilla @Evesy this follows our standard approach, borrowed from the SDK, which is to test for standard environment variables, followed by a listening Managed Identity endpoint as advertised by the MSI_ENDPOINT environment variable, and finally user credentials from developer tools like Azure CLI, VS Code, Azure PowerShell via MSAL secure credential store. There is one slight tweak here as we use the existence of MSI_ENDPOINT environment variable or AZURE_AUTH_METHOD being set to "msi", to enable us to default to Azure CLI auth. Whereas the SDK itself defaults to MSI when using [NewAuthorizerFromEnvironment()](https://github.com/Azure/azure-sdk-for-go/blob/ef9744da754d0cf00d0cfeae7d1a83f2245a4b1c/services/keyvault/auth/auth.go#L26).

Many thanks,
Aaron
